src/Symfony/Component/String/Inflector/EnglishInflector.php

```diff
@@ -0,0 +1,490 @@
+        'atad',
+        'reed',
+        'kcabdeef',
+        'hsif',
+        'ofni',
+        'esoom',
+        'seires',
+        'peehs',
+        'seiceps',
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function singularize(AbstractString $string)
+    {
+        $pluralRev = $string->reverse();
+        $lowerPluralRev = $pluralRev->lower();
+
+        // Check if the word is one which is not inflected, return early if so
+        if (\in_array($lowerPluralRev->toString(), self::$uninflected, true)) {
+            return clone $string;
+        }
+
+        $pluralLength = $lowerPluralRev->length();
+
+        // The outer loop iterates over the entries of the plural table
+        // The inner loop $j iterates over the characters of the plural suffix
+        // in the plural table to compare them with the characters of the actual
+        // given plural suffix
+        foreach (self::$pluralMap as $map) {
+            $suffix = $map[0];
+            $suffixLength = $map[1];
+            $j = 0;
+
+            // Compare characters in the plural table and of the suffix of the
+            // given plural one by one
+            while ($suffix[$j] === $lowerPluralRev->slice($j, 1)->toString()) {
+                // Let $j point to the next character
+                ++$j;
+
+                // Successfully compared the last character
+                // Add an entry with the singular suffix to the singular array
+                if ($j === $suffixLength) {
+                    // Is there any character preceding the suffix in the plural string?
+                    if ($j < $pluralLength) {
+                        $nextIsVocal = false !== strpos('aeiou', $lowerPluralRev->slice($j, 1)->toString());
+
+                        if (!$map[2] && $nextIsVocal) {
+                            // suffix may not succeed a vocal but next char is one
+                            break;
+                        }
+
+                        if (!$map[3] && !$nextIsVocal) {
+                            // suffix may not succeed a consonant but next char is one
+                            break;
+                        }
+                    }
+
+                    $newBase = $string->slice(0, $pluralLength - $suffixLength)->toString();
+                    $newSuffix = $map[4];
+
+                    // Check whether the first character in the plural suffix
+                    // is uppercased. If yes, uppercase the first character in
+                    // the singular suffix too
+                    $firstUpper = ($s = $pluralRev->slice($j - 1, 1))->toString() === $s->upper()->toString();
+
+                    $class = \get_class($string);
+
+                    if (\is_array($newSuffix)) {
+                        $singulars = [];
+
+                        foreach ($newSuffix as $newSuffixEntry) {
+                            $singulars[] = new $class($newBase.($firstUpper ? ucfirst($newSuffixEntry) : $newSuffixEntry));
```

Comment:
using `ucfirst` is not the proper way to deal with unicode strings.

Reply:
`$newSuffixEntry` comes from our own array that has 0 unicode strings.

Reply:
but then, is there any benefit using Unicode-aware manipulation methods elsewhere for things matched against our array ?

Reply:
No and if there are cases, they are my mistakes 😄